### PR TITLE
move select virtual services logic to sidecar

### PR DESCRIPTION
**Pull request by ramaraochavali**
Merges selectVirtualServices logic in http route to IstioEgressListeners selectVirtualServices
Few more minor refactors

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@hzxuzhonghu can you PTAL?

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 1]
```diff
@@ -256,7 +215,7 @@ func BuildSidecarOutboundVirtualHosts(node *model.Proxy, push *model.PushContext
 		if listenerPort == 0 {
 			// Take all ports when listen port is 0 (http_proxy or uds)
 			// Expect virtualServices to resolve to right port
-			servicesByName[svc.Hostname] = svc
+			servicesByName[svc.Hostname] = svc.DeepCopy() // To ensure further changes does not affect service registry.
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:218]: why?

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 2]
```diff
@@ -291,6 +245,38 @@ func BuildSidecarOutboundVirtualHosts(node *model.Proxy, push *model.PushContext
 		}
 	}
 
+	buildVirtualHost := func(hostname string, vhwrapper istio_route.VirtualHostWrapper, svc *model.Service) *route.VirtualHost {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:248]: seems unrelated change

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 3]
```diff
@@ -271,11 +230,6 @@ func BuildSidecarOutboundVirtualHosts(node *model.Proxy, push *model.PushContext
 		}
 	}
 
-	// This is hack to keep consistent with previous behavior.
-	if listenerPort != 80 {
-		// only select virtualServices that matches a service
-		virtualServices = selectVirtualServices(virtualServices, servicesByName)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:277]: `IstioEgressListenerWrapper.VirtualServices` is not equal to this selection, this is port related

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:218]: This is not needed for this PR. But in general added it fo consistency with else block. If you think it is not needed, 
I will remove it

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:248]: As i mentioned in PR - I did few more refactoring to simplify the code

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:277]: I added that logic inside filterVirtualServices

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go [hunk 4]
```diff
@@ -862,7 +862,6 @@ func TestSidecarOutboundHTTPRouteConfig(t *testing.T) {
 				"test-private.com:80": {
 					"test-private.com": true, "test-private.com:80": true, "9.9.9.9": true, "9.9.9.9:80": true,
 				},
-				"match-no-service.not-default:80": {"match-no-service.not-default": true, "match-no-service.not-default:80": true},
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: @hzxuzhonghu This test case correctly fails when we select virtual services with matching services

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: Why do we want this to change? I don't think a VS is required to match a service?

**Review by howardjohn — COMMENTED** [review]
What are the impacts on the generated XDS configuration here? Its hard to grok 600 line change, want to make sure this is not impacting users

**Comment by ramaraochavali** [comment]
> What are the impacts on the generated XDS configuration here? Its hard to grok 600 line change, want to make sure this is not impacting users

This should not impact any XDS config change. This is a fair point. Let me move plain refactorings to a different PR so that it is easy to review this PR. Sorry about that.

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 4 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: https://github.com/istio/istio/pull/34247 - This change if VS host does not find a service, we drop it.

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:277]: https://github.com/istio/istio/blob/77366e7a2b077d426df043669f09fe23d4a22c7e/pilot/pkg/networking/core/v1alpha3/httproute.go#L254-L277 

You can see here it selects the services that has specific port, but i do not see there is port related logic in sidecar

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
As i said they are not equal, TestSelectVirtualService should be fixed

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: The link is just to this PR, was it meant to be somewhere else?

I am a bit concerned we are making a backwards incompatible change.

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: https://github.com/istio/istio/pull/33764 - Sorry this PR has changed that.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: Is that part actually merged?

> If you have concern of breaking some compatibility, we can leave 1, and probihit 2.

seems like it wasn't? Or I do not understand it :slightly_smiling_face:

File: pilot/pkg/model/sidecar.go [hunk 5]
```diff
@@ -634,6 +642,50 @@ func (ilw *IstioEgressListenerWrapper) selectVirtualServices(virtualServices []c
 	return importedVirtualServices
 }
 
+// filterVirtualServices selects the virtual services by matching given services' host names.
+func (ilw *IstioEgressListenerWrapper) filterVirtualServices(virtualServices []config.Config, servicesByName map[host.Name]*Service) []config.Config {
+	// This is hack to keep consistent with previous behavior.
+	if ilw.IstioListener != nil && ilw.IstioListener.Port != nil && ilw.IstioListener.Port.Number == 80 {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:648]: yes, this is not enough. Say we want to select virtualservice for route `8080`, first we need to get only the services that has port `8080`. And then select the vs according to the services that has 8080.

But in this patch, it selects the vses by all the services, am i misreading?

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:648]: That is what this patch does. This selects services based on egress listener config and then apply this filter logic. Please check one more time - If I am missing some flow.

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 4 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: I am confused now. @hzxuzhonghu Is n't this expected if we select VS that only have services in the registry? Basically your fix in #33674 ?

File: pilot/pkg/model/sidecar.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:648]: Maybe i missed， can you point me where the service port is matched? I mean not this hacky 80 port

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:648]: https://github.com/istio/istio/blob/b30f8ab30ab35e6e5608d2c407a17230d6d4bf65/pilot/pkg/model/sidecar.go#L679 - This is an existing code. Egresslistener matching services by port and adds them

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:865]: As i said for route 80, it is hacky, and there is a integration test that depend on the behavior. So i kept it.

File: pilot/pkg/model/sidecar.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:648]: Not sure I understand how this is related to my question? Think about when there is a default sidecar.

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:648]: I think there is some confusion here. This does not change the behaviour of your patch. Your fix of selectVirtualServices is getting services and virtualServices from egressListenerWrapper only https://github.com/istio/istio/blob/bd9ecccaaa83b0fd6a388855d17b313b0f19b627/pilot/pkg/networking/core/v1alpha3/httproute.go#L242 - This patch also is doing the same. Are some thing is missed? I am not able to understand the difference.

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:648]: I see what you mean. Let me fix that

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 4 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go]: Ok. I fixed and this has not changed

**Comment by ramaraochavali** [comment]
@hzxuzhonghu fixed PTAL

**Comment by hzxuzhonghu** [comment]
Thanks for the fix, will be on my way home. Will take a look tomorrow

File: pilot/pkg/model/sidecar.go [hunk 6]
```diff
@@ -634,6 +636,46 @@ func (ilw *IstioEgressListenerWrapper) selectVirtualServices(virtualServices []c
 	return importedVirtualServices
 }
 
+// filterVirtualServices selects the virtual services by matching given services' host names.
+func (ilw *IstioEgressListenerWrapper) filterVirtualServices(virtualServices []config.Config, servicesByName map[host.Name]*Service) []config.Config {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go:640]: @ramaraochavali This is still a separate function from that one.

Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:640]: Yeah. I am wondering how to combine this? Because it is dependent on servicesByPort chosen for that particular RDS. Any suggestions?

**Comment by ramaraochavali** [comment]
@hzxuzhonghu This should be ready. Now I refactored a bit so that we do select virtual services as part of Listener wrapper creation - not in the same `selectVirtualServices` method because after virtual services are selected we add few more services for matching vs destination. Now this is done once per listener conversion rather than every route gen. PTAL

**Comment by hzxuzhonghu** [comment]
I am thinking of a more simple way. Could show you tomorrow, hope it works

**Comment by ramaraochavali** [comment]
@hzxuzhonghu changed. PTAL

File: pilot/pkg/model/sidecar.go [hunk 7]
```diff
@@ -457,20 +456,17 @@ func convertIstioListenerToWrapper(ps *PushContext, configNamespace string,
...
-	out.virtualServices = out.selectVirtualServices(vses, out.listenerHosts)
+	out.virtualServices = SelectVirtualServices(vses, out.listenerHosts)
 	svces := ps.Services(&dummyNode)
 	out.services = out.selectServices(svces, configNamespace, out.listenerHosts)
 
 	return out
 }
 
-// Services returns the list of services imported across all egress listeners by this
-// Sidecar config
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go:468]: why remove

File: pilot/pkg/model/virtualservice_test.go [hunk 8]
```diff
@@ -18,17 +18,26 @@ import (
...
 	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/protocol"
 	"istio.io/istio/pkg/config/schema/collections"
 	"istio.io/istio/pkg/config/visibility"
 )
 
+const wildcardIP = "0.0.0.0"
+
+var tnow = time.Now()
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/virtualservice_test.go:39]: nit：move down to the function

**Review by hzxuzhonghu — COMMENTED** [review]
lgtm, will approve later
